lib/remap-lang/src/type_def.rs

```diff
@@ -206,6 +206,23 @@ impl TypeDef {
         self
     }
 
+    /// Applies a type constraint to the items in an array. If you need all items in the array to
+    /// be integers, for example, set `Kind::Integer`; if items can be either integers or Booleans,
+    /// set `Kind::Integer | Kind::Boolean`; and so on.
+    pub fn fallible_unless_array_has_inner_type(mut self, kind: impl Into<value::Kind>) -> Self {
+        match &self.inner_type_def {
+            Some(InnerTypeDef::Array(inner_kind)) if kind.into() != inner_kind.kind => {
+                self.fallible = true;
+            }
+            None => {
+                self.fallible = true;
+            }
+            _ => (),
```

Comment:
Sorry, I hate to be that irritating buzz in your ear that constantly nags you every step you take! But, we can also have `InnerTypeDef::Both` which represents the situation where something could be either an Array or a Map - in can happen in code like:

```coffeescript
.x = if .blah { [1,2,3] } else { { "foo": 52 } }
```

`Both` should also be fallible.

Thinking about it, the only time it shouldn't be fallible is when the array has the given inner type. So something like this might be simpler:

```rust
match &self.inner_type_def {
        Some(InnerTypeDef::Array(inner_kind)) if kind.into() == inner_kind.kind => (),
        _ => self.fallible = true;
}
```

Reply:
Yeah, I was playing around trying to make the match statement exhaustive going case by case and this was the conclusion I came to as well. Don't ever hesitate to be the irritating buzz when you're right! 🐝